### PR TITLE
Initial master-port of #54039 (support Vault kv version2)

### LINKED ISSUE #50186: BUG: Vault with KV Secrets Engine - Version 2 breaks SDB Vault
### Description of Issue/Question
I am using Vault with KV Secrets Engine - Version 2. This breaks things with SDB since the response from vault is different than KV Secrets Engine - Version 1.

### Setup
Make sure you have Vault configured and enable KV Secrets Engine - Version 2:

```
vault secrets enable -version=2 kv
vault kv enable-versioning secret/
vault kv patch secret/salt/pillar_data wew=lad
```

/etc/salt/master.d/vault.conf:
```
sdb_vault:
  driver: vault

vault:
  url: https://vault_host:8200
  verify: False
  auth:
      method: token
      token: wewladwewladwewlad
  policies:
      - saltstack/minions
      - saltstack/minions/{minion}
```

This patches the sdb.get function to work correctly. I haven't looked into the other sdb functions yet.
```
diff --git a/salt/sdb/vault.py b/salt/sdb/vault.py
index 5980543..81fedbb 100644
--- a/salt/sdb/vault.py
+++ b/salt/sdb/vault.py
@@ -108,7 +108,9 @@ def get(key, profile=None):
             response.raise_for_status()
         data = response.json()['data']

-        return data[key]
+        # If you're using KV Secrets Engine - Version 2
+        # The data is inside the 'data' key
+        return data[key] if data.get(key) else data['data'][key]
     except Exception as e:
         log.error('Failed to read secret! %s: %s', type(e).__name__, e)
         raise salt.exceptions.CommandExecutionError(e)
```

Here's what the response looks like from Vault with kv2, from the data = response.json()['data'] in /opt/saltstack/lib/python2.7/site-packages/salt/sdb/vault.py:get()
```
{u'data': {u'wew': u'lad',}, u'metadata': {u'created_time': u'2018-10-23T20:21:55.042755098Z', u'destroyed': False, u'version': 13, u'deletion_time': u''}}
```

You'll get an error like the following when running the command below:
```
$ sudo salt-run sdb.get 'sdb://sdb_vault/secret/data/salt/pillar_data/wew'
[ERROR   ] Failed to read secret! KeyError: u'wew'
Exception occurred in runner sdb.get: Traceback (most recent call last):
  File "/opt/saltstack/lib/python2.7/site-packages/salt/client/mixins.py", line 387, in _low
    data['return'] = self.functions[fun](*args, **kwargs)
  File "/opt/saltstack/lib/python2.7/site-packages/salt/runners/sdb.py", line 27, in get
    return salt.utils.sdb.sdb_get(uri, __opts__, __utils__)
  File "/opt/saltstack/lib/python2.7/site-packages/salt/utils/sdb.py", line 46, in sdb_get
    return loaded_db[fun](query, profile=profile)
  File "/opt/saltstack/lib/python2.7/site-packages/salt/sdb/vault.py", line 122, in get
    raise salt.exceptions.CommandExecutionError(e)
CommandExecutionError: u'wew'
[INFO    ] Runner completed: 20181023165316837868
```

### Steps to Reproduce Issue
sudo salt-run sdb.get 'sdb://sdb_vault/secret/data/salt/pillar_data/wew'

### Versions Report
Salt Version:
           Salt: 2018.3.2-1735-g84e2dcf

Dependency Versions:
           cffi: 1.11.5
       cherrypy: Not Installed
       dateutil: 2.7.3
      docker-py: 3.5.0
          gitdb: 2.0.4
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.10
        libgit2: 0.27.4
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.5.6
   mysql-python: Not Installed
      pycparser: 2.18
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: 0.27.2
         Python: 2.7.15 (default, Aug 24 2018, 16:24:40)
   python-gnupg: 0.4.3
         PyYAML: 3.13
          PyZMQ: 17.1.2
           RAET: Not Installed
          smmap: 2.0.4
        timelib: Not Installed
        Tornado: 4.5.3
            ZMQ: 4.2.5

System Versions:
           dist: centos 7.5.1804 Core
         locale: UTF-8
        machine: x86_64
        release: 3.10.0-862.11.6.el7.x86_64
         system: Linux
        version: CentOS Linux 7.5.1804 Core

### PR DESCRIPTION
**Pull request by mchugh19**
### What does this PR do?
Attempts to port #54039 to master while adding a unit test for the v2 functionality
This is @thenoid's work to allow salt to support both version 1 and version 2 of Vault's kv secrets backend. 

### What issues does this PR fix or reference?
See notes on #54039
Also fixes #50186

### Previous vs New Behavior
Salt's Vault support really only handled the v1 kv interface. @thenoid's PR added logic to support the correct URLs for both v1 and v2 kv backends.

Vault also changed the return from v1 and v2 kv backends. In v1, secrets were returned. In v2, a data structure is returned like the following:
```
{'data':
  {'data': 
    {'akey': 'avalue',},
  'metadata':
    {'created_time': '2018-10-23T20:21:55.042755098Z',
     'destroyed': False,
     'version': 13,
     'deletion_time': ''}
  }
}
```

The secret data that `vault.read_secret` used to return is now behind another `['data']` lookup. Since this was a change vault-side, this PR does not attempt to retain the old output (maybe users want some of the new data?), and instead publishes what was supplied by vault. 

Thus for users on the v1 kv, running a command like `vault.read_secret 'secret/super/secret` returns a result like: `akey: avalue`
While users of the v2 kv backend running the same command will get a result like displayed above and will need to walk the returned 'data' key to get the same result. 
```{% set supersecret = salt['vault'].read_secret('secret/my/secret')['data'] %}``` (note the trailing 'data')

That said, the execution module's read_secret, and the sdb module's get, both support filtering on the key of the vault secret. So this PR also extends those functions to support the v1 and v2 returns to work as expected when a key is specified. 

### Tests written?
Yes
~~There should really be some docs on the docker setup. I tried to run the integration tests locally so I could investigate updating the vault image, and also running a version 2 kv backend for awesome actual tests, but I'm not sure where the vault docker image is being pulled from.~~ 
Figured it out! Added module and sdb integration tests with the current version of vault's docker image as it's needed for kv v2 support.

### CODE REVIEW

**Comment by thenoid** [comment]
Thank you for doing the needful and for crediting. <3

**Comment by thenoid** [comment]
@Ch3LL ???

File: salt/modules/vault.py [hunk 1]
```diff
@@ -155,7 +159,21 @@ def read_secret(path, key=None):
...
+    Support has been added for Vault KV backend version 2, where the Vault server
+    output has changed to include metadata about the secret. When using KV backend
+    version 2, secret data is now within an additional data key, which will need
+    to be handled in state lookups. For example:
+
+    .. code-block:: jinja
+
+        {% set mysecret = salt['vault'].read_secret('secret/my/secret')['data'] %}
+
+    .. versionchanged:: Neon
```
Comment (Ch3LL) [inline salt/modules/vault.py]: this will need to be updated to Sodium

File: salt/modules/vault.py [hunk 2]
```diff
@@ -243,6 +275,39 @@ def delete_secret(path):
         return False
 
 
+def destroy_secret(path, *args):
+    '''
+    Destory specified secret version at the path in vault. The vault policy
+    used must allow this. Only supported on Vault KV version 2
+
+    .. versionadded:: neon
```
Comment (Ch3LL) [inline salt/modules/vault.py]: Sodium here as well

File: tests/integration/modules/test_vault.py [hunk 3]
```diff
@@ -98,3 +99,121 @@ def test_delete_secret(self):
...
+        '''
+        def count_tests(funcobj):
+            return inspect.ismethod(funcobj) and funcobj.__name__.startswith('test_')
+        numtests = len(inspect.getmembers(VaultTestCaseCurrent, predicate=count_tests))
+        if self.count >= numtests:
+            self.run_state('docker_container.stopped', name='vault')
+            self.run_state('docker_container.absent', name='vault')
+            self.run_state('docker_image.absent', name='vault', force=True)
+
+    @flaky
```
Comment (Ch3LL) [inline tests/integration/modules/test_vault.py]: we are trying not to use this flaky decorator anymore. I know you will see it throughout our code, but we do have future plans to go through and clean up these tests so I don't want to add another flaky test to this list. These flaky tests also cause a lot of maintenance as they still do fail sometimes. Is this test indeed failing sometimes? If so is there something we could do to figure out why it is and make the test more dependable?

File: salt/sdb/vault.py [hunk 4]
```diff
@@ -60,14 +60,18 @@ def set_(key, value, profile=None):
...
+        path = version2['data']
+        data = {'data': data}
 
     try:
         url = 'v1/{0}'.format(path)
-        data = {key: value}
         response = __utils__['vault.make_request'](
             'POST',
             url,
-            profile,
```
Comment (Ch3LL) [inline salt/sdb/vault.py]: can you elaborate why we are removing profile here?

**Review by Ch3LL — CHANGES_REQUESTED** [review]

File: salt/sdb/vault.py (hunk 4 above)
Comment (mchugh19) [reply salt/sdb/vault.py:70]: It is unused in the code. If that's a standard kwarg I'm happy to leave it in.

File: tests/integration/modules/test_vault.py (hunk 3 above)
Comment (mchugh19) [reply tests/integration/modules/test_vault.py:159]: These tests came with the flaky decorator by the time I looked at this. There **shouldn't** be anything particularly flaky about the operations tested. The integration test launches a pre-baked docker image, and runs some http calls against it. As long as the docker states are stable, and the hosting platform is able to run docker images okay, I think the flaky bit can be removed.

File: salt/modules/vault.py (hunk 2 above)
Comment (mchugh19) [reply salt/modules/vault.py]: Done

File: salt/modules/vault.py (hunk 1 above)
Comment (mchugh19) [reply salt/modules/vault.py]: Done

File: tests/integration/modules/test_vault.py (hunk 3 above)
Comment (mchugh19) [reply tests/integration/modules/test_vault.py]: Removed flaky'ness

**Comment by mchugh19** [comment]
Before merging, it's probably worth discussing the topic brought up in the description. Today with vault kv v1,  `vault.read_secret 'secret/super/secret` returns a result like: `akey: avalue`
While users of the v2 kv backend running the same command will get a result like displayed above and will need to walk the returned 'data' key to get the same result.
{% set supersecret = salt['vault'].read_secret('secret/my/secret')['data'] %} (note the trailing 'data')

This is done because vault changed the output when using kv v2 to 
```
{'data':
  {'data': 
    {'akey': 'avalue',},
  'metadata':
    {'created_time': '2018-10-23T20:21:55.042755098Z',
     'destroyed': False,
     'version': 13,
     'deletion_time': ''}
  }
}
```
Where there are separated data and metadata keys. Maybe it makes sense to return data by default, and to have a toggle on metadata? 
`vault.read_secret /secret/my/path metadata=True`

**Comment by thenoid** [comment]
So with `metadata=True` it'd return kv2 style {'data': {whatever}, 'metadata': {something}}?

`metadata=False` it'd be a kv1 style?

I think that's a fair abstraction

> Before merging, it's probably worth discussing the topic brought up in the description. Today with vault kv v1, `vault.read_secret 'secret/super/secret` returns a result like: `akey: avalue`
> While users of the v2 kv backend running the same command will get a result like displayed above and will need to walk the returned 'data' key to get the same result.
> {% set supersecret = salt['vault'].read_secret('secret/my/secret')['data'] %} (note the trailing 'data')
> 
> This is done because vault changed the output when using kv v2 to
> 
> ```
> {'data':
>   {'data': 
>     {'akey': 'avalue',},
>   'metadata':
>     {'created_time': '2018-10-23T20:21:55.042755098Z',
>      'destroyed': False,
>      'version': 13,
>      'deletion_time': ''}
>   }
> }
> ```
> 
> Where there are separated data and metadata keys. Maybe it makes sense to return data by default, and to have a toggle on metadata?
> `vault.read_secret /secret/my/path metadata=True`

**Comment by mchugh19** [comment]
As described in the comments, the previous behavior of vault.read_secret has been retained. Now read_secret will return the secret data on both kv v1 and v2. Anyone needing to access the full return, can use `vault.read_secret /secret/path metadata=True`

**Review by dwoz — APPROVED** [review]

**Pull request merged**